primitives/runtime/src/traits.rs

```diff
@@ -1210,13 +1213,29 @@ pub trait Header: Clone + Send + Sync + Codec + Eq + MaybeSerialize + Debug + 's
 	}
 }
 
+/// Something that provides the Header Type.
+pub trait HeaderProvider {
+	/// Header type.
+	type HeaderT: Header<Hash = Self::HashT>;
+	/// Block hash type.
+	type HashT: HashOutput;
+}
```

Comment:
We should add some docs above this on why we do this (workaround for a compiler bug) and that this will be removed at some point.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=92e282e4441562b05c37225b931c89d9 does this maybe also work? If yes, we could use this.

Reply:
For the naming I would go with `Header` and `Hash`.

Reply:
And why do we need `Hash` here? As `Hash` is declared in `Header`.

Reply:
> For the naming I would go with Header and Hash.

That was showing a conflict with `Header` and `Hash` in `Block` requiring a cast at each usage to avoid disambiguity.

> And why do we need Hash here? As Hash is declared in Header.

I can try removing that as well, but a lot of places use `Block` to retrieve the Hash. Would need a change in those as well. Possibly a follow-up?

Reply:
>  https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=92e282e4441562b05c37225b931c89d9 does this maybe also work? If yes, we could use this.

Super interesting! Let me try this as well. Thank you :)

Reply:
> That was showing a conflict with `Header` and `Hash` in `Block` requiring a cast at each usage to avoid disambiguity.

Ahh yeah, I didn't thought about this case! Makes sense. 



> I can try removing that as well, but a lot of places use `Block` to retrieve the Hash. Would need a change in those as well. Possibly a follow-up?

I meant removing `Hash` from the provider trait, not from `Block`. And then everywhere we fetch it from the provider, we go over the header.

Reply:
> We should add some docs above this on why we do this (workaround for a compiler bug) and that this will be removed at some point.

Added

> https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=92e282e4441562b05c37225b931c89d9 does this maybe also work? If yes, we could use this.

Tried this, but still causes the same error.